### PR TITLE
21585 codegen logger

### PR DESCRIPTION
**Pull request by lancegliser**

### CODE REVIEW

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
+  // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+
+  // List of extensions which should be recommended for users of this workspace.
+  "recommendations": ["ronnidc.nunjucks"],
```
Comment (lancegliser) [inline .vscode/extensions.json:6]: Is the one you're using?

File: src/types/shims/Train.ts [hunk 2]
```diff
@@ -2,4 +2,5 @@ export type TrainResponse = {
   ablated_indices?: number[];
   num_trained: number;
   status?: "analyze" | "analyzed" | null;
+  warnings: string[][];
```
Comment (fulpm) [inline src/types/shims/Train.ts:5]: This should not be part of the body of the train response

File: src/client/AbstractBaseClient.ts [hunk 3]
```diff
@@ -20,8 +21,20 @@ export type ExecuteResponse<R = unknown> = {
   warnings: string[];
 };
 
+export type AbstractBaseClientOptions = {
+  /** Enable tracing all Trainee operations for debugging. */
+  trace?: boolean;
```
Comment (fulpm) [inline src/client/AbstractBaseClient.ts:26]: Trace is not necessarily available for turning on/off

File: src/client/worker/HowsoWorkerClient.ts [hunk 4]
```diff
@@ -548,27 +551,33 @@ export class HowsoWorkerClient extends AbstractBaseClient {
...
+          }
           size = yield;
         }
       }.bind(this),
       batchOptions,
     );
 
     await this.autoPersistTrainee(trainee.id);
-    return { num_trained, status, ablated_indices };
+    return { num_trained, status, ablated_indices, warnings };
```
Comment (fulpm) [inline src/client/worker/HowsoWorkerClient.ts:581]: Warnings should not be returned in the Train data.

File: src/client/worker/HowsoWorkerClient.ts [hunk 5]
```diff
@@ -548,27 +551,33 @@ export class HowsoWorkerClient extends AbstractBaseClient {
     let num_trained = 0;
     let status = null;
     const ablated_indices: number[] = [];
+    const warnings: string[][] = [];
```
Comment (fulpm) [inline src/client/worker/HowsoWorkerClient.ts:554]: I understand using a 2d array gets you the warnings per batch, but not sure it makes sense to give all the duplicate warnings back.

File: src/types/shims/Train.ts (hunk 2 above)
Comment (lancegliser) [reply src/types/shims/Train.ts:5]: Oh, right. That's the _batch_ train response. Probably need to create an additional response type. Sec.

File: src/client/worker/HowsoWorkerClient.ts (hunk 4 above)
Comment (lancegliser) [reply src/client/worker/HowsoWorkerClient.ts:581]: I have to disagree here. There's no other place for accessing the produced warnings if we don't return them. Did you have another plan?

File: src/client/AbstractBaseClient.ts (hunk 3 above)
Comment (lancegliser) [reply src/client/AbstractBaseClient.ts]: Moved trace out to the specific client options.

File: src/client/worker/HowsoWorkerClient.ts (hunk 4 above)
Comment (lancegliser) [reply src/client/worker/HowsoWorkerClient.ts]: Introduced ClientBatchResponse standard per our call.

**Pull request merged**